src/tools/tidy/src/error_codes.rs

```diff
@@ -0,0 +1,397 @@
+                path.display()
+            ));
+            return;
+        }
+
+        // Make sure that the file is referenced in `error_codes.rs`
+        let filename = path.file_name().unwrap().to_str().unwrap().split_once('.');
+        let err_code = filename.unwrap().0; // `unwrap` is ok because we know the filename is in the correct format.
+
+        if error_codes.iter().all(|e| e != err_code) {
+            errors.push(format!(
+                "Found valid file `{}` in error code docs directory without corresponding \
+                entry in `error_code.rs`",
+                path.display()
+            ));
+            return;
+        }
+
+        // `has_test.0` checks whether the error code has any (potentially untested) code example.
+        // `has_test.1` checks whether the error code has a proper (definitely tested) doctest.
+        let has_test = check_explanation_has_doctest(&contents, &err_code);
+        if has_test.2 {
+            emit_ignore_warning += 1;
+        }
+        if has_test.3 {
+            no_longer_emitted_codes.push(err_code.to_owned());
+            emit_no_longer_warning += 1;
+        }
+        if !has_test.0 {
+            emit_no_code_warning += 1;
+        }
+
+        let test_ignored = IGNORE_DOCTEST_CHECK.contains(&err_code);
+
+        // Check that the explanation has a doctest, and if it shouldn't, that it doesn't
+        if !has_test.1 && !test_ignored {
+            errors.push(format!(
+                "`{}` doesn't use its own error code in compile_fail example",
+                path.display(),
+            ));
+        } else if has_test.1 && test_ignored {
+            errors.push(format!(
+                "`{}` has a compile_fail doctest with its own error code, it shouldn't \
+                be listed in `IGNORE_DOCTEST_CHECK`",
+                path.display(),
+            ));
+        }
+    });
+
+    if emit_ignore_warning > 0 {
+        println!(
+            "WARNING: {emit_ignore_warning} error codes use the ignore header. This should not be used, add the error codes to the \
+            `IGNORE_DOCTEST_CHECK` constant instead. This *will* become a hard error."
+        );
+    }
+    if emit_no_code_warning > 0 {
+        println!(
+            "WARNING: {emit_ignore_warning} error codes don't have a code example, all error codes are expected \
+            to have one (even if untested). This *will* become a hard error."
+        );
+    }
+    if emit_no_longer_warning > 0 {
+        println!(
+            "WARNING: {emit_no_longer_warning} error codes are no longer emitted and should be removed entirely. \
+            This *will* become a hard error."
+        );
+    }
+
+    no_longer_emitted_codes
+}
+
+/// This function returns a tuple indicating whether the provided explanation:
+/// a) has a code example, tested or not.
+/// b) has a valid doctest
+fn check_explanation_has_doctest(explanation: &str, err_code: &str) -> (bool, bool, bool, bool) {
```

Comment:
Functions like this that return a bunch of booleans are error prone and not self-documenting. 

An alternative would be to return a struct, or return a tuple of boolean-like enums, e.g. 
```rust
enum HasDocTest{
    Yes,
    No
}

enum HasUiTest{
    Yes,
    No
}
// etc
```

or:

```rust
struct TestChecks{
    has_ui: bool,
    has_doc: bool,
    // ...
}
```

Reply:
As below, I can change this if you want, but these aren't permanent and just exist until I fix up the warnings and remove half of that function.

Reply:
Somewhat fixed with the `--verbose` changes and code simplification in that regard. This will be further improved soon when these warnings are removed.